File: pkg/src/wilson_hilferty.c
/* $ID: wilson_hilferty.c, last updated 2024-01-03, F.Osorio */

#include "fastmatrix.h"

/* Wilson-Hilferty transformation */

void
wilson_hilferty_chisq(double *distances, int *n, int *p, double *z)
{ /* Wilson-Hilferty transformation for chi-squared variables (to be called by R) */
  FM_WH_chisq(distances, *n, *p, z);
}

void
wilson_hilferty_gamma(double *y, int *n, double *shape, double *scale, double *z)
{ /* Wilson-Hilferty transformation for gamma variables (to be called by R) */
  FM_WH_gamma(y, *n, *shape, *scale, z);
}

void
FM_WH_gamma(double *y, int n, double shape, double scale, double *z)
{ /* Wilson-Hilferty transformation for gamma variables */
  double t, mean, sd, q = 1./3.;

  for (int i = 0; i < n; i++) {
    t = *y++ / (shape / scale);
    mean = 1. - 1. / (9. * shape);
    sd = 1. / sqrt(9. * shape);
    *z++ = (R_pow(t, q) - mean) / sd;
  }
}

void
FM_WH_chisq(double *distances, int n, int p, double *z)
{ /* Wilson-Hilferty transformation for chi-squared variables */
  double f, q = 1./3., s = 2./9.;

  for (int i = 0; i < n; i++) {
    f = *distances++ / p;
    *z++ = (R_pow(f, q) - (1. - s / p)) / sqrt(s / p);
  }
}

void
FM_WH_F(double *distances, int n, int p, double eta, double *z)
{ /* Wilson-Hilferty transformation for F variables */
  double f, q = 1./3., r = 2./3., s = 2./9.;

  for (int i = 0; i < n; i++) {
    f  = *distances++ / p;
    f /= 1. - 2. * eta;
    *z++ = ((1. - s * eta) * R_pow(f, q) - (1. - s / p)) / sqrt(s * eta * R_pow(f, r) + s / p);
  }
}

void
FM_WH_Laplace(double *distances, int n, int p, double *z)
{ /* Wilson-Hilferty transformation for Laplace variables */
  double f, mean, sd, q = 1./3.;

  for (int i = 0; i < n; i++) {
    f = *distances++ / (2. * p);
    mean = 1. - 1. / (9. * p);
    sd = 1. / sqrt(9. * p);
    *z++ = (R_pow(f, q) - mean) / sd;
  }
}

File: pkg/R/wilson_hilferty.R
## ID: wilson_hilferty.R, last updated 2024-01-03, F.Osorio

wilson.hilferty <- function(x, shape, rate = 1)
{ ## Wilson-Hilferty transformation for gamma deviates
  if (!is.vector(x))
    stop("'x' must be a vector")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  if (any(x < 0)) 
    stop("'x' must be non-negative")
  n <- length(x)

  if (is.null(shape))
    stop("'shape' must be provided")
  if (shape < 0) 
    stop("'shape' must be non-negative")
  if (rate < 0) 
    stop("'rate' must be non-negative")

  z <- .C("wilson_hilferty_gamma",
          x = as.double(x),
          n = as.integer(n),
          shape = as.double(shape),
          rate = as.double(rate),
          z = double(n))$z
  z
}

WH.normal <- function(x)
{ ## Wilson-Hilferty transformation for multivariate normal deviates
  if (is.data.frame(x))
    x <- as.matrix(x)
  else if (!is.matrix(x))
    stop("'x' must be a matrix or a data frame")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  n <- nrow(x)
  p <- ncol(x)

  z <- cov.weighted(x, weights = rep(1, n))
  D2 <- Mahalanobis(x, center = z$mean, cov = z$cov, inverted = FALSE)
  z <- .C("wilson_hilferty_chisq",
          distances = as.double(D2),
          n = as.integer(n),
          p = as.integer(p),
          z = double(n))$z
  z
}
